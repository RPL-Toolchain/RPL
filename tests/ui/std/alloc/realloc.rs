//@revisions: normal inline
//@[normal]compile-flags: -Z inline-mir=false
//@[inline]compile-flags: -Z inline-mir=true
//@[inline]check-pass
use std::alloc::{Layout, alloc, dealloc, realloc};

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn use_after_realloc_u8() {
    let layout = Layout::array::<u8>(16).unwrap();
    unsafe {
        let ptr = alloc(layout) as *mut u8;
        assert!(!ptr.is_null());
        ptr.write(42);
        let new_ptr = realloc(ptr as *mut u8, layout, layout.size() * 2) as *mut u8;
        assert!(!new_ptr.is_null());
        // use the old pointer after realloc
        let x = *ptr;
        assert_eq!(x, 42);
        dealloc(ptr as *mut u8, layout);
    }
}

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn use_after_realloc_u32() {
    let layout = Layout::array::<u32>(16).unwrap();
    unsafe {
        let ptr = alloc(layout) as *mut u32;
        assert!(!ptr.is_null());
        ptr.write(42);
        let new_ptr = realloc(ptr as *mut u8, layout, layout.size() * 2) as *mut u32;
        assert!(!new_ptr.is_null());
        // use the old pointer after realloc
        let x = *ptr;
        //~[normal]^ERROR: use a pointer from `u32` after it's reallocated
        assert_eq!(x, 42);
        dealloc(ptr as *mut u8, layout);
    }
}
