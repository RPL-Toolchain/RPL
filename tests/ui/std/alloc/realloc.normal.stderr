error: use a pointer from `u32` after it's reallocated
  --> tests/ui/std/alloc/realloc.rs:33:17
   |
LL |         let new_ptr = realloc(ptr as *mut u8, layout, layout.size() * 2) as *mut u32;
   |                       -------------------------------------------------- `realloc` called here
...
LL |         let x = *ptr;
   |                 ^^^^ used here
   |
   = note: the `u32` buffer may be reallocated here
   = note: `-D rpl::use-after-realloc` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::use_after_realloc)]`

error: aborting due to 1 previous error

